### PR TITLE
(NETTY-11) Added BodyProducer support

### PR DESCRIPTION
**Pull request by chtyim**
- Allow sending data asynchronously
  - Driven by output channel availability.

### CODE REVIEW

File: src/main/java/co/cask/http/BasicHttpResponder.java [hunk 1]
```diff
@@ -130,6 +141,120 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+      public void operationComplete(ChannelFuture future) throws Exception {
+        if (!future.isSuccess()) {
+          callBodyProducerHandleError(bodyProducer, future.getCause());
+          channel.close();
+          return;
+        }
+
+        try {
+          HttpChunk chunk = new DefaultHttpChunk(bodyProducer.nextChunk());
+          size += chunk.getContent().readableBytes();
```
Comment (poornachandra) [inline src/main/java/co/cask/http/BasicHttpResponder.java]: We can compute the size only when `contentLength >= 0`

File: src/main/java/co/cask/http/BasicHttpResponder.java [hunk 2]
```diff
@@ -130,6 +141,120 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+      @Override
+      public void operationComplete(ChannelFuture future) throws Exception {
+        if (!future.isSuccess()) {
+          callBodyProducerHandleError(bodyProducer, future.getCause());
+          channel.close();
+          return;
+        }
+
+        try {
+          bodyProducer.finished();
```
Comment (poornachandra) [inline src/main/java/co/cask/http/BasicHttpResponder.java:246]: When `bodyProducer.finished()` throws exception, channel won't be closed here.

**Comment by poornachandra** [comment]
@chtyim Two comments, please fix and merge.

File: src/main/java/co/cask/http/BasicHttpResponder.java (hunk 2 above)
Comment (chtyim) [inline src/main/java/co/cask/http/BasicHttpResponder.java:246]: it's close in the catch block.

Comment (poornachandra) [inline src/main/java/co/cask/http/BasicHttpResponder.java:246]: Right, missed that.

File: src/main/java/co/cask/http/BasicHttpResponder.java (hunk 1 above)
Comment (chtyim) [inline src/main/java/co/cask/http/BasicHttpResponder.java]: Fixed

**Comment by chtyim** [comment]
Squashed and merging.

**Pull request merged**